Introduction to android SafeBrowsing API and virus Total Api (#1655)
OWASP/owasp-mstg: Document/0x05h-Testing-Platform-Interaction.md

The [Vulnerable App](https://github.com/clviper/android-fragment-injection/raw/master/vulnerableapp.apk "Vulnerable App Fragment Injection") and [Exploit PoC App](https://github.com/clviper/android-fragment-injection/blob/master/exploit.apk "PoC App to exploit Fragment Injection") are available for downloading.

> ### Testing for Safe Browsing in WebViews (MSTG-PLATFORM-2)

Reviewer: After writing all the other comments, I'm tending to think that if we're simply testing for SafeBrowsing as the Android SafeBrowsing feature, that wouldn't require a full section. However, if we're also including restrictions to a trusted domain things get more interesting and Android SafeBrowsing could be also included as an enhancement for a safer browsing experience. In that case, we might need another title for this section, I'm not sure yet.

Reviewer 2: I think we can keep this section. It also talks about VirusTotal and we can extend it in the future with the point you raised about trusted domains. 
But we should change the title to "Testing URL Loading in WebViews (MSTG-PLATFORM-2)".

#### Overview

WebViews are Android's embedded components which allow you're app to open web pages within your application. In addition to mobile apps realated threats, WebViews expose your app to known web threats (e.g, XSS,OpenRedirct, etc).

> To give you control over the pages loaded by your WebView, Android provides `shouldOverrideUrlLoading` methods which allows your application to either abort loading WebViews with suspicious content by returning `true` or allow the WebView load the URL by returnig `false`.

Reviewer: ```suggestion
To give you control over the pages loaded by your WebView, Android provides `shouldOverrideUrlLoading` methods which allows your application to either abort loading WebViews with suspicious content by returning `true` or allow the WebView to load the URL by returning `false`.
```

Reviewer: I think this part is worth further research. What happens if a developer decides to use `shouldOverrideUrlLoading`? Is it triggered for any resources loaded from the page itself? e.g. images, scripts, etc. Is it triggered by HTTP requests made by JavaScript code on the page?

We're testing for Safe browsing, how can we be sure that the app will be only loading content from trusted sources?

Next, related to the questions above, does it make sense to **add a Dynamic Analysis section** (in order to test the effectiveness of the approach)?
- What if the page I'm accessing is in the domain of the app but it loads resources from other untrusted sources?
- What if the domain validation isn't sufficient? e.g. the app only verifies that the domain ends with domain.com but will accept evil.domain.com?

@commjoen @sushi2k @TheDauntless any thoughts here?

Author: Thanks @cpholguera for the detailed review. as far as i know 'shouldOverrideUrlLoading' can be used by the developer whenever a URL is beeing loaded in a WebView, and it's up to the developer whether to load the URL or not. This means that it allows the developer to restrict WebViews to load specific domains using this method to prevent any unknown or unpredicted domain from being loaded.

Furthermore i think that is possible for the developer to check other request's parameters than only the domain before loading the URL, and by other parameters i mean specially HTTP headers (e.g, Access-Control-Allow-Origin ) or other security headers.

Reviewer 2: According to the documentation this method doesn't work for POST requests.  This is something we should note here. https://developer.android.com/reference/android/webkit/WebViewClient#shouldOverrideUrlLoading(android.webkit.WebView,%20android.webkit.WebResourceRequest)

Let's mention that here and make a link to the documentation. Then I think this gives a good overview for now and is pointing the reader in the right direction. 

Answering Carlos questions and analysing this function would then be a little (and interesting) research project on it's own :-) We can create a ticket for it to follow-up.

Reviewer 2: https://github.com/OWASP/owasp-mstg/issues/1678

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Reviewer: If I understand SafeBrowsing properly, there's no option to restrict the app to a trusted domain, right? It will be enabled by default and the only thing to check is if it's disabled explicitly by the developer. Even if used as an API, there's also no way to restrict the app to a trusted domain. It's rather protecting against phishing or whatever is considered a threat by Google. It's making the browsing safer but not necessarily secure (as the name nicely indicates). Therefore, I'm not sure if we should give so many details or keep it shorter letting people know that it's a nice to have but it's not necessarily increasing the security. That they should rather focus on potential URL/domain validation errors that may happen when trying to restrict the app to a trusted domain (?). See the other big comment above :) 

Any thoughts here as well guys? @commjoen @sushi2k @TheDauntless

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Author: I join you on what you said about SafeBrowsing. SafeBrowsing is not based on whitelisting principal, which means it will not allow the developer to restrict the app to a trusted domain, but it's rather based on blacklisting principal which allow the developer to protect the app from loading a URL blacklisted by Google for being a potential threat.

This section is not meant to present SafeBrowsing as an ultimate solution for securing WebView, but rather an addition layer of protection to profit from this Android feature and make browsing safer. Implementing multiple layers of protection is always recommanded for more security, and mixing SafeBrowsing with URL/domain validation and other technics is always recommanded.

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Reviewer: Definitely. I think this really goes in line with this other comment: https://github.com/OWASP/owasp-mstg/pull/1655#discussion_r384149782

We should then give it a re-think :)

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Author: I Suggest that we use a different name for the section and add more details about other technics (Android/Java related) for protecting WebViews. the section will still cover MSTG-PLATEFORM-2.

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Reviewer 2: Carlos already brought up the title "Testing URL Loading in WebViews (MSTG-PLATFORM-2)", I think this title makes sense, as it sounds more neutral.

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Reviewer 2: > I Suggest that we use a different name for the section and add more details about other technics (Android/Java related) for protecting WebViews. the section will still cover MSTG-PLATEFORM-2.

@A-AFTAHI  We also have testcases for MSTG-PLATFORM-5, MSTG-PLATFORM-6 and MSTG-PLATFORM-7 which are all about hardening WebViews. What exactly would you like to add to MSTG-PLATFORM-2 which is not covered in the other test cases? Just want to avoid redundant content.

> For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Author: @sushi2k each of The MSTG-PLATFORM-5, MSTG-PLATFORM-6 and MSTG-PLATFORM-7 covers a specific topic : (Javascript usage, protocol handlers then native methods) whereas MSTG-PLATFORM-2 is more general and covers input validation and santitization so SafeBrowsing does fit MSTG-PLATFORM-2 and not the others.

In other words if not adding SafeBrowsing within MSTG-PLATFORM-2 we'll have to add a new chapter for it because it will not make sens to be part of the other ones mentioned before.

By default, WebViews show a warning to users about the security risk with the option to load the URL or stop the page from loading. With the SafeBrowsing API you can customize your application's behavior by either reporting the threat to SafeBrowsing or performing a particular action such as returning back to safety each time it encounters a known threat. Please check the [Android Developers documentation](https://developer.android.com/about/versions/oreo/android-8.1#safebrowsing) for usage examples.

> You can use the SafeBrowsing API independently from WebViews using the [SafetyNet library](https://developer.android.com/training/safetynet/safebrowsing), which implements a client for Safe Browsing Network Protocol v4. SafetyNet allows you to analyze all the URLs that your app is supposed load. You can check URLs with different schemes (e.g : http,file,etc) since SafeBrowsing is agnostic to URL schemes, and against `TYPE_POTENTIALLY_HARMFUL_APPLICATION` and `TYPE_SOCIAL_ENGINEERING` threat types.

Reviewer: ```suggestion
You can use the SafeBrowsing API independently from WebViews using the [SafetyNet library](https://developer.android.com/training/safetynet/safebrowsing), which implements a client for Safe Browsing Network Protocol v4. SafetyNet allows you to analyze all the URLs that your app is supposed load. You can check URLs with different schemes (e.g. http, file) since SafeBrowsing is agnostic to URL schemes, and against `TYPE_POTENTIALLY_HARMFUL_APPLICATION` and `TYPE_SOCIAL_ENGINEERING` threat types.
```

Devices running platforms older than Android 4.4 (API level 19) use a version of WebKit that has several security issues. As a workaround, the app must confirm that WebView objects [display only trusted content](https://developer.android.com/training/articles/security-tips.html#WebView "WebView Best Practices") if the app runs on these devices.

> For more secure web browsing, Android 8.1 introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

Reviewer: ```suggestion
For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.
```

> For more secure web browsing, Android 8.1 introduces an implementation for [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4). SafeBrowsing allows your application to detect URLs that Google has classified as a known threat.

Reviewer 3: ```suggestion
For more secure web browsing, Android 8.1 introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.
```

> By default the webview shows a warning to users about the security risk with the option to load the URL or return to the safe page. With the SafeBrowsing API you can customize your application behavior by either reporting the threat to SafeBrowsing or performing a particular action such as returning back to safety each time it encounter a known threat.

Reviewer 3: ```suggestion
By default the WebView shows a warning to users about the security risk with the option to load the URL or stop the page from loading. With the SafeBrowsing API you can customize your application's behavior by either reporting the threat to SafeBrowsing or performing a particular action such as returning back to safety each time it encounters a known threat.
```

```Java
public class MyWebViewClient extends WebViewClientCompat {
    // Automatically go "back to safety" when attempting to load a website that
    // Google has identified as a known threat. An instance of WebView calls
    // this method only after Safe Browsing is initialized, so there's no
    // conditional logic needed here.
    @Override
    public void onSafeBrowsingHit(WebView view, WebResourceRequest request,
            int threatType, SafeBrowsingResponseCompat callback) {

>         // The "true" argument indicates that your app reports incidents like

Reviewer 3: Move these comments down so that they are above the callback.backToSafety line

Reviewer: Please do so :)

Reviewer: I've checked and this example is the same as in Android Developers. I'd suggest to link to it directly and remove it from here for now. You can put something like: please check the [Android Developers documentation](link "title") for usage examples.

Author: @cpholguera @TheDauntless if you have no more suggestions i guess we could move the content to the MSTG-PLATFORM-2 section. i'll use the same branch for that.

Reviewer: Hi @A-AFTAHI , thanks for the changes, please take a look at this message below:https://github.com/OWASP/owasp-mstg/pull/1655#pullrequestreview-361087479

        // this one to Safe Browsing.
        if (WebViewFeature.isFeatureSupported(WebViewFeature.SAFE_BROWSING_RESPONSE_BACK_TO_SAFETY)) {
            callback.backToSafety(true);
            Toast.makeText(view.getContext(), "Unsafe web page blocked.",
                    Toast.LENGTH_LONG).show();
        }
    }
}
```

> [Virus Total API](https://support.virustotal.com/hc/en-us/articles/115002146469-API-Scripts) also can be used for analyzing and checking URLs and local files for known threats. various implementaions of scripts to communicate with the API has been publically shared by the community.

Reviewer 3: ```suggestion
[Virus Total API](https://support.virustotal.com/hc/en-us/articles/115002146469-API-Scripts) also provides an API for analyzing URLs and local files for known threats. A Java implementation for the API is available on Virus Total's documentation page.
```

> ##### When sending URLs or files to be checked for know threats make sure they don't contain sensitive data which could abuse users privacy, or expose your app for security breaches.

Reviewer 3: ```suggestion
##### When sending URLs or files to be checked for know threats make sure they don't contain sensitive data which could compromise a user's privacy, or expose sensitive content from your application.
```

Reviewer: I guess this is meant to be a note, right? Please put `> ` instead of `###..`

Devices running platforms older than Android 4.4 (API level 19) use a version of WebKit that has several security issues. As a workaround, the app must confirm that WebView objects [display only trusted content](https://developer.android.com/training/articles/security-tips.html#WebView "WebView Best Practices") if the app runs on these devices.

For more secure web browsing, Android 8.1 (API level 27) introduces the [`SafeBrowsing API`](https://developers.google.com/safe-browsing/v4), which allows your application to detect URLs that Google has classified as a known threat.

> By default the WebView shows a warning to users about the security risk with the option to load the URL or stop the page from loading. With the SafeBrowsing API you can customize your application's behavior by either reporting the threat to SafeBrowsing or performing a particular action such as returning back to safety each time it encounters a known threat.

Reviewer: New line
```suggestion

By default, WebViews show a warning to users about the security risk with the option to load the URL or stop the page from loading. With the SafeBrowsing API you can customize your application's behavior by either reporting the threat to SafeBrowsing or performing a particular action such as returning back to safety each time it encounters a known threat.
```

```Java
public class MyWebViewClient extends WebViewClientCompat {
    // Automatically go "back to safety" when attempting to load a website that
    // Google has identified as a known threat. An instance of WebView calls
    // this method only after Safe Browsing is initialized, so there's no
    // conditional logic needed here.
    @Override
    public void onSafeBrowsingHit(WebView view, WebResourceRequest request,
            int threatType, SafeBrowsingResponseCompat callback) {
        // The "true" argument indicates that your app reports incidents like
        // this one to Safe Browsing.
        if (WebViewFeature.isFeatureSupported(WebViewFeature.SAFE_BROWSING_RESPONSE_BACK_TO_SAFETY)) {
            callback.backToSafety(true);
            Toast.makeText(view.getContext(), "Unsafe web page blocked.",
                    Toast.LENGTH_LONG).show();
        }
    }
}
```

> You can use the SafeBrowsing API seperatly from WebViews using the [SafetyNet library](https://developer.android.com/training/safetynet/safebrowsing), which implements a client for Safe Browsing Network Protocol v4. SafetyNet allows you to analyze all the URLs that your app is supposed load. You can check URLs with different schemes againt multiple threat types.

Author: the SafeBrowsing API is agnostic to url schemes which means that https://www.google.com & www.google.com are the same for it. In other words different schemes (ex: http,https,file...) are supported.

SafeBrowsing API check currently for two threat types (https://developers.google.com/android/reference/com/google/android/gms/safetynet/SafeBrowsingThreat.html#getThreatType() ) : 

- TYPE_POTENTIALLY_HARMFUL_APPLICATION

- TYPE_SOCIAL_ENGINEERING
I will add this details to the PR if required
